ext4: mark inode dirty after initializing the extent tree

From: Aneesh Kumar K.V <aneesh.kumar@linux.vnet.ibm.com>

We should mark inode dirty only after initializing the extent
tree. Also if we fail during extent initialization we need
to call DQUOT_FREE_INODE.

Signed-off-by: Aneesh Kumar K.V <aneesh.kumar@linux.vnet.ibm.com>
---

 fs/ext4/ialloc.c |   13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)


Index: linux-2.6.25-rc8/fs/ext4/ialloc.c
===================================================================
--- linux-2.6.25-rc8.orig/fs/ext4/ialloc.c	2008-04-09 17:45:36.000000000 -0700
+++ linux-2.6.25-rc8/fs/ext4/ialloc.c	2008-04-10 07:27:02.000000000 -0700
@@ -835,11 +835,6 @@ got:
 	if (err)
 		goto fail_free_drop;
 
-	err = ext4_mark_inode_dirty(handle, inode);
-	if (err) {
-		ext4_std_error(sb, err);
-		goto fail_free_drop;
-	}
 	if (test_opt(sb, EXTENTS)) {
 		/* set extent flag only for diretory, file and normal symlink*/
 		if (S_ISDIR(mode) || S_ISREG(mode) || S_ISLNK(mode)) {
@@ -848,10 +843,16 @@ got:
 			err = ext4_update_incompat_feature(handle, sb,
 					EXT4_FEATURE_INCOMPAT_EXTENTS);
 			if (err)
-				goto fail;
+				goto fail_free_drop;
 		}
 	}
 
+	err = ext4_mark_inode_dirty(handle, inode);
+	if (err) {
+		ext4_std_error(sb, err);
+		goto fail_free_drop;
+	}
+
 	ext4_debug("allocating inode %lu\n", inode->i_ino);
 	goto really_out;
 fail:
